### DISCUSSION TITLE
Streamline Hotkeys

### DISCUSSION THREAD

**Discussion by butterw** [triplea-game/triplea#9315] [Feature Requests & Ideas]
## Streamline Hotkeys

From a development perspective, putting all current shortcuts as constants in a single code file is simple enough to do and has many advantages:
- Developpers would know what all the current hotkeys are (+ their history) and could change them easily.
- Advanced Users could change hotkeys by recompiling with a custom shortcut file.
- The current hotkeys could be displayed automatically in a table, which would guarantee up-to-date info for users.
- Eventually all hotkeys may become user configurable in the GUI or a config file.

I've looked into it a bit, my suggestion is that all keyboard hotkeys should be defined as constants with a meaningfull name in a single file (ex: org.triplea.swing.key.hotkeys). 
Instead of using harcoded values, import the hotkeys file for all menu accelerators/keyboard listeners.  
KeyEvent.VK_RIGHT >> KeyCode.RIGHT >> HotKeys.SCROLL_RIGHT1 

> What would be the correct naming and data type to use in Java, 
Is it simply a matter of renaming the keys in KeyCode ? 
Hotkeys are KeyEvent/KeyStrokes (the result of a combination of key). 

```
public final class UrlConstants {
  public static final String AXIS_AND_ALLIES_FORUM = "https://www.axisandallies.org/forums";

import java.awt.event.KeyEvent;
public enum KeyCode {
  SPACE(KeyEvent.VK_SPACE),
```

ex in TripleAFrame.java: 
if (keyCode == KeyEvent.VK_RIGHT)
would become 	
if (keyCode == HotKeys.ScrollRight1)
In this case we can have multiple hotkeys for the same action (to be able to support WASD or equivalent)
- KeyEvent are integer, a value of -1 would mean the key is disabled

in most cases there would be a single hotkey for an action.
ex: newSaveMenu with a value KeyCode.S
- menu accelerators are used in combination with CTRL or ("meta" if Mac)
- the modification isn't required for menu mnemonics right now. 
- See also #9296

**Comment by butterw** [comment]
- Keyboard hotkeys are typically accelerators (on Actions or Menuitems) or frame keybindings.
These hotkeys are KeyStrokes. They can be defined via the overloaded method KeyStroke.getKeyStroke

1) Moving default keyboard hotkeys to a single HotKey class is achievable. See https://github.com/butterw/ui
2) User configurable hotkeys can then be configured via a .properties file (Keystrokes can be defined with a String, with the correct syntax).
- They can override default hotkeys.
3) All hotkeys (user configured or default) could be put in a map object <String hotKeyName, Keystroke>. This would allows displaying all current hotkeys.
4) Changing hotkeys at runtime would likely be troublesome.

**Reply by DanVanAtta** [reply]
Seeing and being able to configure all hotkeys in one place would be very sweet. Runtime changes is certainly desirable.
